### PR TITLE
networkd: Tunnel add support to configure key for VTI/VTI6

### LINKED ISSUE #3298: Implement key/mark for VTI/VTI6 tunnel interfaces
### Submission type
- [ ] Bug report
- [x] Request for enhancement (RFE)

Implement key/mark for VTI/VTI6 tunnel interfaces.

See commit: a613382bbf4357ce13f17c988713b80172e091fb

### PR DESCRIPTION
**Pull request by ssahani**
fixes #3298

```
[NetDev]                                                                                                                                                                                      
Name=vti-tun                                                                                                                                                                                  
Kind=vti                                                                                                                                                                                      
MTUBytes=1480                                                                                                                                                                                 

[Tunnel]                                                                                                                                                                                      
Local=10.65.208.212                                                                                                                                                                           
Remote=10.65.208.213                                                                                                                                                                          
#Key=10.65.208.213                                                                                                                                                                            
InputKey=111                                                                                                                                                                                  
OutputKey=222 
```

```
$ ip tunnel show 
ip_vti0: ip/ip remote any local any ttl inherit nopmtudisc key 0
vti-tun: ip/ip remote 10.65.208.213 local 10.65.208.212 dev enp0s25 ttl inherit nopmtudisc ikey 111 okey 222
```

### CODE REVIEW

File: man/systemd.netdev.xml [hunk 1]
```diff
@@ -639,6 +639,28 @@
         </listitem>
       </varlistentry>
       <varlistentry>
+        <term><varname>Key=</varname></term>
+        <listitem>
+          <para>The <literal>Key</literal> parameter specifies the same key to use in
+          both directions (InputKey and OutputKey). The <literal>Key</literal> is either a
```
Comment (poettering) [inline man/systemd.netdev.xml]: always write options with the "=" suffix, and enclose in `<varname>` please. i.e. change here:

`<literal>Key</literal>` → `<varname>Key=</varname>`
`InputKey` → `<varname>InputKey=</varname>`

File: man/systemd.netdev.xml [hunk 2]
```diff
@@ -639,6 +639,28 @@
...
+          number or an IPv4 address-like dotted quad. It is only used for VTI/VTI6 tunnels.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>InputKey=</varname></term>
+        <listitem>
+          <para>The <literal>InputKey</literal> parameter specifies the key to use for input.
+          The format is same as <literal>Key</literal>. It is only used for VTI/VTI6 tunnels.</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: similar here

File: man/systemd.netdev.xml [hunk 3]
```diff
@@ -639,6 +639,28 @@
...
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>OutputKey=</varname></term>
+        <listitem>
+          <para>The <literal>OutputKey</literal> parameter specifies the key to use for output.
+          The format is same as <literal>Key</literal>. It is only used for VTI/VTI6 tunnels.</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: and here...

File: man/systemd.netdev.xml (hunk 1 above)
Comment (poettering) [inline man/systemd.netdev.xml]: also, can you elaborate on what a "key" is in this context?

File: src/network/networkd-netdev-tunnel.c [hunk 4]
```diff
@@ -202,6 +202,7 @@ static int netdev_ip6gre_fill_message_create(NetDev *netdev, Link *link, sd_netl
 
 static int netdev_vti_fill_message_create(NetDev *netdev, Link *link, sd_netlink_message *m) {
         Tunnel *t = VTI(netdev);
+        unsigned ikey, okey;
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: shouldn't this be uint32_t? You use sd_netlink_message_append_u32() to add it to th enetlink msg...

File: src/network/networkd-netdev-tunnel.h [hunk 5]
```diff
@@ -48,6 +48,9 @@ typedef struct Tunnel {
         unsigned ttl;
         unsigned tos;
         unsigned flags;
+        unsigned key;
+        unsigned ikey;
+        unsigned okey;
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.h]: shouldn't this be uint32_t?

File: src/network/networkd-netdev-tunnel.c [hunk 6]
```diff
@@ -214,6 +215,21 @@ static int netdev_vti_fill_message_create(NetDev *netdev, Link *link, sd_netlink
         if (r < 0)
                 return log_netdev_error_errno(netdev, r, "Could not append IFLA_IPTUN_LINK attribute: %m");
 
+        if (t->key)
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: please don't use C's downgrade-to-bool feature except when checking for ptrs and actual bools. i.e. given that `t->key` is actually numeric, please change the check to `if (t->key != 0)` or so.

File: src/network/networkd-netdev-tunnel.c [hunk 7]
```diff
@@ -239,6 +256,22 @@ static int netdev_vti6_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+
+        r = sd_netlink_message_append_u32(m, IFLA_VTI_IKEY, ikey);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append IFLA_VTI_IKEY attribute: %m");
+
+        r = sd_netlink_message_append_u32(m, IFLA_VTI_OKEY, okey);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append IFLA_VTI_OKEY attribute: %m");
+
+
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c:276]: spurious newline

Comment (poettering) [inline src/network/networkd-netdev-tunnel.c:276]: given that the lines above are used twice, maybe make a funciton out of them?

File: src/network/networkd-netdev-tunnel.c [hunk 8]
```diff
@@ -413,6 +446,50 @@ int config_parse_tunnel_address(const char *unit,
...
+        Tunnel *t = userdata;
+        unsigned k;
+        int r, f;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = in_addr_from_string_auto(rvalue, &f, &buffer);
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: shouldn't this be `in_addr_from_string(AF_INET, …)` instead? this cannot be ipv6, can it?

File: src/network/networkd-netdev-tunnel.c [hunk 9]
```diff
@@ -413,6 +446,50 @@ int config_parse_tunnel_address(const char *unit,
...
+                            const char *section,
+                            unsigned section_line,
+                            const char *lvalue,
+                            int ltype,
+                            const char *rvalue,
+                            void *data,
+                            void *userdata) {
+        union in_addr_union buffer;
+        Tunnel *t = userdata;
+        unsigned k;
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: shouldn't this be uint32_t?

File: src/network/networkd-netdev-tunnel.c [hunk 10]
```diff
@@ -413,6 +446,50 @@ int config_parse_tunnel_address(const char *unit,
...
+        int r, f;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = in_addr_from_string_auto(rvalue, &f, &buffer);
+        if (r < 0) {
+                r = safe_atou(rvalue, &k);
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: shouldn't this be safe_atou32?

File: man/systemd.netdev.xml [hunk 11]
```diff
@@ -639,6 +639,30 @@
         </listitem>
       </varlistentry>
       <varlistentry>
+        <term><varname>Key=</varname></term>
+        <listitem>
+          <para>The <literal>Key=</literal> parameter specifies the same key to use in
```
Comment (poettering) [inline man/systemd.netdev.xml]: `<varname>` not `<literal>` for field names please. (here and everywhere else)

File: man/systemd.netdev.xml [hunk 12]
```diff
@@ -639,6 +639,30 @@
...
+          both directions (<literal>InputKey=</literal> and <literal>OutputKey=</literal>).
+          The <literal>Key=</literal> is either a number or an IPv4 address-like dotted quad.
+          It is used as mark-configured SAD/SPD entry as part of the lookup key (both in data
+          and control path) in ip xfrm. It is only used for VTI/VTI6 tunnels.</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: what is "ip xfrm", can you find better wording for that?

File: src/network/networkd-netdev-tunnel.c [hunk 13]
```diff
@@ -413,6 +448,45 @@ int config_parse_tunnel_address(const char *unit,
...
+        assert(data);
+
+        r = in_addr_from_string(AF_INET, rvalue, &buffer);
+        if (r < 0) {
+                r = safe_atou32(rvalue, &k);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Failed to parse tunnel key ignoring assignment: %s", rvalue);
+                        return 0;
+                }
+        }
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: hmm, you parse into "buffer", but never write the result into "k" then.

**Comment by poettering** [comment]
added a couple of more notes. Could you please make sure to test all PRs before submitting them?

File: man/systemd.netdev.xml (hunk 12 above)
Comment (ssahani) [inline man/systemd.netdev.xml]: framework used to implement IPsec protocol

File: src/network/networkd-netdev-tunnel.c (hunk 13 above)
Comment (ssahani) [inline src/network/networkd-netdev-tunnel.c]: ooops I missed it .

File: man/systemd.netdev.xml (hunk 12 above)
Comment (poettering) [inline man/systemd.netdev.xml]: well, if that's the case then this should be in the man page of course, not here in the comments.

Comment (ssahani) [inline man/systemd.netdev.xml]: yes we have actually . http://man7.org/linux/man-pages/man8/ip-xfrm.8.html will add the link here

File: src/network/networkd-netdev-tunnel.c [hunk 14]
```diff
@@ -413,6 +448,46 @@ int config_parse_tunnel_address(const char *unit,
...
+
+        r = in_addr_from_string(AF_INET, rvalue, &buffer);
+        if (r < 0) {
+                r = safe_atou32(rvalue, &k);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Failed to parse tunnel key ignoring assignment: %s", rvalue);
+                        return 0;
+                }
+        } else
+                k = buffer.in.s_addr;
```
Comment (poettering) [inline src/network/networkd-netdev-tunnel.c]: hmm, so s_addr is stored in big endian, but when passing the key later on into the netlink message you use htonl() on the field, thus suggesting native endian. That looks wrong. I figure you need an invocation of ntohl() here?

(btw, htonl() is an awful name for what it does, consider using htobe32() which is equivalent but much more descriptive...)

**Comment by halvors** [comment]
See issue #4371

**Pull request merged**